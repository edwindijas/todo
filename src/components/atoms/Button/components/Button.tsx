interface ButtonProps {
  children: React.ReactElement | string,
  className: string,
  onClick?: () => void,
  dataTestId: string
}

 const Button = ({
  className,
  children,
  onClick,
  dataTestId
}: ButtonProps): JSX.Element =>
  <button data-testid={dataTestId} onClick={onClick} className={className}>
    {children}
  </button>


export default Button