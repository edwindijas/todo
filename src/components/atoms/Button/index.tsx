import { cva, type VariantProps } from 'class-variance-authority'
import CusButton from './components/Button'
import CustomLink from './components/Link'

const Button = ({
  className = '',
  to = '',
  isLink = false,
  isNavLink = false,
  children,
  startIcon,
  endIcon,
  dataTestId = 'button-btn',
  onClick,
  ...variants
}: ButtonProps): JSX.Element => {
  const Component = isLink || isNavLink ? CustomLink : CusButton

  return <Component to={to}
    isNavLink={isNavLink}
    onClick={onClick}
    dataTestId={dataTestId}
    className={`${className} ${buttonVariants(variants)}`} >
      <>
        {startIcon !== undefined &&
        <div className={iconVariants({left: true, ...variants})}>
          {startIcon}
        </div>
        }
        <span className='w-full' >{children}</span>
        { endIcon !== undefined &&
          <div>
            {endIcon}
          </div>
        }
      </>
  </Component>
}

const buttonVariants = cva(['flex w-full capitalize items-center gap-2 relative'], {
  variants: {
    intent: {
      none: '',
      primary: 'text-center text-white bg-blue-medium',
      secondary: 'text-center border border-blue-medium bg-blue-light text-blue-medium fill-blue-medium'
    },
    padding: {
      sm: 'px-1',
      md: 'px-2',
      lg: 'px-4'
    },
    size: {
      lg: '',
      md: 'h-14',
      sm: ''
    },
    isNavLink: {
      true: '',
      false: ''
    },
    rounded: {
      'none': '',
      'sm': 'rounded-sm',
      'md': 'rounded-md',
      'lg': 'rounded-lg',
      'full': 'rounded-full'
    }
  },
  defaultVariants: {
    intent: 'none',
    size: 'md',
    padding: 'md',
    rounded: 'md'
  }
})

const iconVariants = cva('shrink-0', {
  variants: {
    left: {
      true: '',
      false: ''
    },
    size: {
      lg: '',
      md: 'h-6 w-6',
      sm: ''
    },
    float: {
      true: 'absolute top-0 bottom-0 my-auto'
    }
  },
  compoundVariants: [
    {
      float: true,
      left: true,
      className: 'left-4',
    }
  ],
  defaultVariants: {
    size: 'md'
  }
})

export type ButtonVariantProps = VariantProps<typeof buttonVariants>;
export type IconVariantProps = VariantProps<typeof iconVariants>

interface ButtonProps extends ButtonVariantProps, IconVariantProps {
  className?: string,
  isLink?: boolean
  isNavLink?: boolean,
  to?: string
  children: React.ReactElement | string
  onClick?: () => void,
  startIcon?: React.ReactElement,
  endIcon?: React.ReactElement,
  dataTestId?: string
}

export default Button
