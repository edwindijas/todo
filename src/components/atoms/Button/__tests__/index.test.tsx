import { fireEvent } from '@testing-library/react'
import Button from '..'
import render from '@/utils/test-utils/customRender'

const onClick = jest.fn((): void => {})

describe ('button', () => {
  it ('matches snapshot', () => {
    const {container} = render(<>
      <Button isLink isNavLink className='someGEneric' >Button 1</Button>
      <Button isNavLink className='someGeneric' >Button 1</Button>
      <Button isLink >Button 3</Button>
      <Button >Button 1</Button>
      <Button className='someGEneric' >Button 1</Button>
    </>)

    expect(container).toMatchSnapshot()
  })

  it ('expects normal button', () => {
    const {queryByRole} = render(<Button >Button1</Button>)
    expect(queryByRole('button')).toBeInTheDocument()
    expect(queryByRole('link')).not.toBeInTheDocument()
  })

  it('correctly renders link', () => {
    const {queryByRole} = render(<Button isLink >Button</Button>)
    expect(queryByRole('link')).toBeInTheDocument()
    expect(queryByRole('button')).not.toBeInTheDocument()
  }) 

  it ('it calls onClick', () => {
    const {getByTestId} = render(<Button onClick={onClick} >Button 1</Button>)
    fireEvent.click(getByTestId('button-btn'))
    expect(onClick).toHaveBeenCalled()
  })
})