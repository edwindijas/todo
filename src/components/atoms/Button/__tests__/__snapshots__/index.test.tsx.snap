// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button matches snapshot 1`] = `
<div>
  <a
    aria-current="page"
    class="someGEneric flex capitalize items-center gap-2 relative px-2 h-12 rounded-md bg-blue-light text-blue-medium fill-blue-medium relative before:absolute before:top-2 before:bottom-2 before:left-0 before:bg-blue-medium before:w-1 before:block before:rounded-sm"
    href="/"
  >
    <span
      class="w-full"
    >
      Button 1
    </span>
  </a>
  <a
    aria-current="page"
    class="someGeneric flex capitalize items-center gap-2 relative px-2 h-12 rounded-md bg-blue-light text-blue-medium fill-blue-medium relative before:absolute before:top-2 before:bottom-2 before:left-0 before:bg-blue-medium before:w-1 before:block before:rounded-sm"
    href="/"
  >
    <span
      class="w-full"
    >
      Button 1
    </span>
  </a>
  <a
    class=" flex capitalize items-center gap-2 relative px-2 h-12 rounded-md"
    href="/"
  >
    <span
      class="w-full"
    >
      Button 3
    </span>
  </a>
  <button
    class=" flex capitalize items-center gap-2 relative px-2 h-12 rounded-md"
    data-testid="button-btn"
  >
    <span
      class="w-full"
    >
      Button 1
    </span>
  </button>
  <button
    class=" flex capitalize items-center gap-2 relative px-2 h-12 rounded-md"
    data-testid="btn-test-id"
  >
    <span
      class="w-full"
    >
      Button 1
    </span>
  </button>
  <button
    class="someGEneric flex capitalize items-center gap-2 relative px-2 h-12 rounded-md"
    data-testid="button-btn"
  >
    <span
      class="w-full"
    >
      Button 1
    </span>
  </button>
</div>
`;
